[macOS] Vertical scrollbar is misplaced when a nonzero left content inset is set

whsieh opened the pull request:
#### b6b2dc97d5c146df11893fc2d18f37177aff696b
<pre>
[macOS] Vertical scrollbar is misplaced when a nonzero left content inset is set
<a href="https://bugs.webkit.org/show_bug.cgi?id=288943">https://bugs.webkit.org/show_bug.cgi?id=288943</a>
<a href="https://rdar.apple.com/145979548">rdar://145979548</a>

Reviewed by Megan Gardner.

Fix logic in `ScrollView::updateScrollbars` to properly account for left/right content insets:

-   When computing the vertical scrollbar frame, don&apos;t offset by the left content inset; this just
    pushes the scrollbar offscreen.

-   When computing the horizontal scrollbar frame, offset it by the left content inset and shrink
    the size to account for both left/right insets. This ensures that the horizontal scrollbar isn&apos;t
    obscured by the left content inset area.

Test: scrollbars/mac/top-left-content-insets-scrollbar-position.html

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.propertiesOfLayerWithID):

Add a `UIHelper` method to asynchronously grab platform layer properties by layer ID.

(window.UIHelper.getCALayerTree):

Drive-by fix: remove an unnecessary early return if `!isIOSFamily()`, so that `getCALayerTree` can
work on macOS as well.

* LayoutTests/scrollbars/mac/top-left-content-insets-scrollbar-position-expected.txt: Added.
* LayoutTests/scrollbars/mac/top-left-content-insets-scrollbar-position.html: Added.

Add a layout test to exercise the change.

* LayoutTests/tiled-drawing/top-left-content-insets.html:

Drive-by fix: try to un-flake this layout test by delaying the test by a presentation update.

* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::updateScrollbars):
* Source/WebCore/testing/Internals.cpp:
(WebCore::getLayerID):
(WebCore::Internals::horizontalScrollbarLayerID const):
(WebCore::Internals::verticalScrollbarLayerID const):

Add internal test helpers to get the platform `GraphicsLayer` IDs for the horizontal and vertical
scrollbars.

* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _propertiesOfLayerWithID:]):

Move this into `WKWebViewTesting.mm`, so that we can use it on both macOS and iOS.

* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _propertiesOfLayerWithID:]): Deleted.
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::propertiesOfLayerWithID const):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::propertiesOfLayerWithID const): Deleted.

Canonical link: <a href="https://commits.webkit.org/291496@main">https://commits.webkit.org/291496@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0349788305e65713a15e747129b603ce892e7cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93047 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12598 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2294 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98046 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43573 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95097 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21052 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71142 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28563 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96049 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9706 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84192 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51471 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9400 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1826 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42886 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79691 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100071 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14738 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80168 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20351 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80090 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79470 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19740 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24015 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1322 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13176 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20083 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25259 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19770 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23230 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21511 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
```diff
@@ -52,6 +52,7 @@ struct WKAppPrivacyReportTestingData {
 
 @property (nonatomic, readonly) NSString *_caLayerTreeAsText;
 
+- (NSDictionary *)_propertiesOfLayerWithID:(unsigned long long)layerID;
```
maybe this can be typed as `NSDictionary<NSString *, id> *` for a tad more type safety?

whsieh committed:
[macOS] Vertical scrollbar is misplaced when a nonzero left content inset is set
https://bugs.webkit.org/show_bug.cgi?id=288943
rdar://145979548

Reviewed by Megan Gardner.

Fix logic in `ScrollView::updateScrollbars` to properly account for left/right content insets:

-   When computing the vertical scrollbar frame, don't offset by the left content inset; this just
    pushes the scrollbar offscreen.

-   When computing the horizontal scrollbar frame, offset it by the left content inset and shrink
    the size to account for both left/right insets. This ensures that the horizontal scrollbar isn't
    obscured by the left content inset area.

Test: scrollbars/mac/top-left-content-insets-scrollbar-position.html

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.propertiesOfLayerWithID):

Add a `UIHelper` method to asynchronously grab platform layer properties by layer ID.

(window.UIHelper.getCALayerTree):

Drive-by fix: remove an unnecessary early return if `!isIOSFamily()`, so that `getCALayerTree` can
work on macOS as well.

* LayoutTests/scrollbars/mac/top-left-content-insets-scrollbar-position-expected.txt: Added.
* LayoutTests/scrollbars/mac/top-left-content-insets-scrollbar-position.html: Added.

Add a layout test to exercise the change.

* LayoutTests/tiled-drawing/top-left-content-insets.html:

Drive-by fix: try to un-flake this layout test by delaying the test by a presentation update.

* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::updateScrollbars):
* Source/WebCore/testing/Internals.cpp:
(WebCore::getLayerID):
(WebCore::Internals::horizontalScrollbarLayerID const):
(WebCore::Internals::verticalScrollbarLayerID const):

Add internal test helpers to get the platform `GraphicsLayer` IDs for the horizontal and vertical
scrollbars.

* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _propertiesOfLayerWithID:]):

Move this into `WKWebViewTesting.mm`, so that we can use it on both macOS and iOS.

* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _propertiesOfLayerWithID:]): Deleted.
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::propertiesOfLayerWithID const):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::propertiesOfLayerWithID const): Deleted.

Canonical link: https://commits.webkit.org/291496@main
